src/sync/sync.nim

```diff
@@ -0,0 +1,238 @@
+        # file in problem-specifications (or `description.md` if that file
+        # doesn't exist).
+        let instrFilename = "instructions.md"
+        let trackInstrPath = trackDocsDir / instrFilename
+        if fileExists(trackInstrPath):
+          let descFilename = "description.md"
+          let psInstrPath = psExerciseDir / instrFilename
+          let psDescPath = psExerciseDir / descFilename
+          if fileExists(psInstrPath):
+            checkFilesIdentical(psInstrPath, trackInstrPath, slug,
+                               instrFilename, seenUnsynced, conf, result)
+          elif fileExists(psDescPath):
+            checkFilesIdentical(psDescPath, trackInstrPath, slug,
+                                instrFilename, seenUnsynced, conf, result)
+          else:
+            logNormal(&"[error] {slug}: does not have an upstream " &
+                      &"{instrFilename} or {descFilename} file")
+            seenUnsynced.incl skDocs
+        else:
+          logNormal(&"[error] {slug}: {instrFilename} is missing")
+          seenUnsynced.incl skDocs
+
+      else:
+        logDetailed(&"[skip] {slug}: does not exist in problem-specifications")
+    else:
+      logNormal(&"[error] {slug}: .docs dir missing")
+      seenUnsynced.incl skDocs
+
+proc checkFilepaths(exercises: seq[Exercise], seenUnsynced: var set[SyncKind]) =
+  if false:
+    seenUnsynced.incl skFilepaths
+
+type
+  MetadataPair = object
+    trackJsonPath: string
+    updatedJson: JsonNode
+
+proc checkMetadata(exercises: seq[Exercise],
+                   psExercisesDir: string,
+                   trackPracticeExercisesDir: string,
+                   seenUnsynced: var set[SyncKind],
+                   conf: Conf): seq[MetadataPair] =
+  for exercise in exercises:
+    let slug = exercise.slug.string
+    let trackMetaDir = joinPath(trackPracticeExercisesDir, slug, ".meta")
+
+    if dirExists(trackMetaDir):
+      let psExerciseDir = psExercisesDir / slug
+      if dirExists(psExerciseDir):
+
+        let metadataFilename = "metadata.toml"
+        let psMetadataTomlPath = psExerciseDir / metadataFilename
+        if fileExists(psMetadataTomlPath):
+          let trackExerciseConfigPath = trackMetaDir / "config.json"
+          if fileExists(trackExerciseConfigPath):
+            let toml = parsetoml.parseFile(psMetadataTomlPath)
+            var j = json.parseFile(trackExerciseConfigPath)
+            const keys = ["blurb", "source", "source_url"]
+            for key in keys:
+              if toml.hasKey(key):
+                let upstreamVal = toml[key]
+                if upstreamVal.kind == TomlValueKind.String:
+                  if j.hasKey(key):
+                    let trackVal = j[key]
+                    if trackVal.kind == JString:
+                      if upstreamVal.stringVal == trackVal.str:
+                        logDetailed(&"[skip] {slug}: metadata is up-to-date")
```

Comment:
Is this message being output three times due to it being called inside the `for` loop?

Reply:
Yes, thanks. Will fix.

Reply:
Fixed. Hopefully.